dance/dance_pipeline.py

```diff
@@ -100,9 +169,236 @@ def _generate_molecules_from_database(self) -> oechem.OEMol:
+        | SMILES       | The name of a file for writing SMILES strings (`str` or `pathlib.Path`) |
+        +--------------+-------------------------------------------------------------------------+
+
+        ``sorted_oeb`` is stored in the ``sorted_by_fingerprint_oeb`` attribute
+        as a ``pathlib.Path`` for future reference by the pipeline.
+
+        Parameters
+        ----------
+        selection_frequency : int
+            How often to select molecules.
+        dataset_type : str
+            The type of dataset.
+        dataset_info : Union[str, pathlib.Path]
+            Info for the dataset -- refer to above table for acceptable types.
+        sorted_oeb : Union[str, pathlib.Path], optional
+            Name of an OEB (Openeye Binary) file for storing the molecules
+            sorted by fingerprint. If this file already exists, it will be
+            overwritten!
+        in_memory_sorting_threshold : int
+            The number of molecules that can be sorted in memory at once. This
+            is a low-level parameter for the sorting algorithm, and most users
+            should not have to deal with it. Note that if this threshold is
+            maximized, then the maximum number of molecules that can be sorted
+            by the pipeline is approximately the square of this threshold.
+
+        Raises
+        ------
+        RuntimeError
+            If the provided ``dataset_type`` is not supported.
+        RuntimeError
+            If ``selection_frequency`` is less than 1.
+        """
+        if dataset_type not in self.SUPPORTED_DATASET_TYPES:
+            raise RuntimeError(f"`{dataset_type}` is not a supported dataset type")
+        if selection_frequency < 1:
+            raise RuntimeError(f"selection_frequency({selection_frequency}) must be an integer greater >= 1")
+
+        self.sorted_by_fingerprint_oeb = pathlib.Path(sorted_oeb)
+        self._sort_molecules_by_fingerprint(self.fingerprint_output_oeb, self.sorted_by_fingerprint_oeb,
+                                            in_memory_sorting_threshold)
+        sorted_molstream = oechem.oemolistream(str(self.sorted_by_fingerprint_oeb))
+
+        # Open any necessary files for outputting the dataset. This will vary by
+        # dataset_type.
+        if dataset_type == "SMILES":
+            dataset_stream = oechem.oemolostream(str(dataset_info))
+            dataset_stream.SetFormat(oechem.OEFormat_SMI)
+
+        # Perform the selection.
+        for idx, mol in enumerate(sorted_molstream.GetOEMols()):
+            if idx % selection_frequency == 0:
+                if dataset_type == "SMILES":
+                    oechem.OEWriteMolecule(dataset_stream, mol)
+
+        # Clean up.
+        if dataset_type == "SMILES":
+            dataset_stream.close()
+
+        sorted_molstream.close()
+
+    @staticmethod
+    def _sort_molecules_by_fingerprint(input_oeb: pathlib.Path,
```

Comment:
I'm guessing this function will be pretty tough to understand. Let me know if you would like me to explain it in a meeting.